unit/java_bytecode/java_replace_nondet/replace_nondet.cpp

```diff
@@ -0,0 +1,120 @@
+/*******************************************************************\
+
+ Module: Tests for the replace nondet pass to make sure it works both
+         when remove_returns has been run before and after the call.
+
+ Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <testing-utils/load_java_class.h>
+
+#include <goto-programs/goto_convert_functions.h>
+#include <goto-programs/remove_virtual_functions.h>
+#include <util/config.h>
+#include <goto-instrument/cover.h>
+#include <util/options.h>
+#include <iostream>
+#include <goto-programs/remove_returns.h>
+#include <goto-programs/replace_java_nondet.h>
+#include <goto-programs/remove_instanceof.h>
+
+void validate_method_removal(
+  std::list<goto_programt::instructiont> instructions)
+{
+  bool method_removed = true, replacement_nondet_exists = false;
+
+  // Quick loop to check that our method call has been replaced.
+  for(const auto &inst : instructions)
+  {
+    if(inst.is_assign())
+    {
+      const code_assignt &assignment = to_code_assign(inst.code);
+      if (assignment.rhs().id() == ID_side_effect)
+      {
+        const side_effect_exprt &see = to_side_effect_expr(assignment.rhs());
+        if(see.get_statement() == ID_nondet)
+        {
+          replacement_nondet_exists = true;
+        }
+      }
+    }
+
+    if(inst.is_function_call())
+    {
+      const code_function_callt &function_call =
+        to_code_function_call(inst.code);
+
+      // Small check to make sure the instruction is a symbol.
+      if(function_call.function().id() != ID_symbol)
+        continue;
+
+      const irep_idt function_id =
+        to_symbol_expr(function_call.function()).get_identifier();
+      if(
+        function_id ==
+        "java::org.cprover.CProver.nondetWithoutNull:()"
+        "Ljava/lang/Object;")
+      {
+        method_removed = false;
+      }
+    }
+  }
+
+  REQUIRE((method_removed && replacement_nondet_exists));
```

Comment:
Not blocking but in future I'd split this into two require statements - then the output will be more useful (it will say which one is not true).

Reply:
Thanks :+1: